Sources/SwiftGRPCNIOInteroperabilityTests/InteroperabilityTestCases.swift

```diff
@@ -0,0 +1,679 @@
+import SwiftGRPCNIO
+import NIOHTTP1
+
+/// This test verifies that implementations support zero-size messages. Ideally, client
+/// implementations would verify that the request and response were zero bytes serialized, but
+/// this is generally prohibitive to perform, so is not required.
+///
+/// Server features:
+/// - EmptyCall
+///
+/// Procedure:
+/// 1. Client calls EmptyCall with the default Empty message
+///
+/// Client asserts:
+/// - call was successful
+/// - response is non-null
+///
+/// It may be possible to use UnaryCall instead of EmptyCall, but it is harder to ensure that the
+/// proto serialized to zero bytes.
+class EmptyUnary: InteroperabilityTest {
+  func run(using connection: GRPCClientConnection) throws {
+    let client = Grpc_Testing_TestServiceService_NIOClient(connection: connection)
+    let call = client.emptyCall(Grpc_Testing_Empty())
+
+    try assertEqual(call.response, Grpc_Testing_Empty())
+    try assertEqual(call.status.map { $0.code }, .ok)
+  }
+}
+
+/// This test verifies that gRPC requests marked as cacheable use GET verb instead of POST, and
+/// that server sets appropriate cache control headers for the response to be cached by a proxy.
+/// This test requires that the server is behind a caching proxy. Use of current timestamp in the
+/// request prevents accidental cache matches left over from previous tests.
+///
+/// Server features:
+/// - CacheableUnaryCall
+///
+/// Procedure:
+/// 1. Client calls CacheableUnaryCall with SimpleRequest request with payload set to current
+///    timestamp. Timestamp format is irrelevant, and resolution is in nanoseconds. Client adds a
+///    x-user-ip header with value 1.2.3.4 to the request. This is done since some proxys such as
+///    GFE will not cache requests from localhost. Client marks the request as cacheable by
+///    setting the cacheable flag in the request context. Longer term this should be driven by
+///    the method option specified in the proto file itself.
+/// 2. Client calls CacheableUnaryCall again immediately with the same request and configuration
+///    as the previous call.
+///
+/// Client asserts:
+/// - Both calls were successful
+/// - The payload body of both responses is the same.
+class CacheableUnary: InteroperabilityTest {
+  func run(using connection: GRPCClientConnection) throws {
+    let client = Grpc_Testing_TestServiceService_NIOClient(connection: connection)
+
+    var timestamp = DispatchTime.now().rawValue
+    let request = Grpc_Testing_SimpleRequest.withPayload(of: .bytes(of: &timestamp))
+
+    var headers = HTTPHeaders()
+    headers.add(name: "x-user-ip", value: "1.2.3.4")
+    let callOptions = CallOptions(customMetadata: headers, cacheable: true)
+
+    let call1 = client.cacheableUnaryCall(request, callOptions: callOptions)
+    let call2 = client.cacheableUnaryCall(request, callOptions: callOptions)
+
+    try assertEqual(call1.response.map { $0.payload }, call2.response.map { $0.payload })
```

Comment:
Should we also compare this to the payload in `request`? Also, given that we are not behind a caching proxy, I guess these two are only identical because the corresponding requests had identical payloads as well?

Reply:
> Should we also compare this to the payload in `request`?

The payload is ignored by the server: 
https://github.com/grpc/grpc-swift/blob/283754e80bb30a630e5e06508af654bc1d68fd7d/Sources/SwiftGRPCNIOInteroperabilityTests/TestServiceProvider_NIO.swift#L89

I'll add a comment to elaborate.

> Also, given that we are not behind a caching proxy, I guess these two are only identical because the corresponding requests had identical payloads as well?

This is just the client end of the test, we won't run this against a NIO server (for now). Some server implementations support this test by faking a caching proxy; this is probably something we can do if we allow additional handlers to be added to the server.